// posthog-js/lib/src/utils/request-router.js
import { document } from '../utils/globals';
import { logger } from './logger';
/**
 * The request router helps simplify the logic to determine which endpoints should be called for which things
 * The basic idea is that for a given region (US or EU), we have a set of endpoints that we should call depending
 * on the type of request (events, replays, decide, etc.) and handle overrides that may come from configs or the decide endpoint
 */
export var RequestRouterRegion;
(function (RequestRouterRegion) {
    RequestRouterRegion["US"] = "us";
    RequestRouterRegion["EU"] = "eu";
    RequestRouterRegion["CUSTOM"] = "custom";
})(RequestRouterRegion || (RequestRouterRegion = {}));
var ingestionDomain = 'i.posthog.com';
var RequestRouter = /** @class */ (function () {
    function RequestRouter(instance) {
        this._regionCache = {};
        this.instance = instance;
    }
    Object.defineProperty(RequestRouter.prototype, "apiHost", {
        get: function () {
            return this.instance.config.api_host.trim().replace(/\/$/, '');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RequestRouter.prototype, "uiHost", {
        get: function () {
            var _a;
            var host = (_a = this.instance.config.ui_host) === null || _a === void 0 ? void 0 : _a.replace(/\/$/, '');
            if (host === 'https://app.posthog.com') {
                return 'https://us.posthog.com';
            }
            return host;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RequestRouter.prototype, "region", {
        get: function () {
            // We don't need to compute this every time so we cache the result
            if (!this._regionCache[this.apiHost]) {
                if (/https:\/\/(app|us|us-assets)(\.i)?\.posthog\.com/i.test(this.apiHost)) {
                    this._regionCache[this.apiHost] = RequestRouterRegion.US;
                }
                else if (/https:\/\/(eu|eu-assets)(\.i)?\.posthog\.com/i.test(this.apiHost)) {
                    this._regionCache[this.apiHost] = RequestRouterRegion.EU;
                }
                else {
                    this._regionCache[this.apiHost] = RequestRouterRegion.CUSTOM;
                }
            }
            return this._regionCache[this.apiHost];
        },
        enumerable: false,
        configurable: true
    });
    RequestRouter.prototype.endpointFor = function (target, path) {
        if (path === void 0) { path = ''; }
        if (path) {
            path = path[0] === '/' ? path : "/".concat(path);
        }
        if (target === 'ui') {
            return (this.uiHost || this.apiHost.replace(".".concat(ingestionDomain), '.posthog.com')) + path;
        }
        if (this.region === RequestRouterRegion.CUSTOM) {
            return this.apiHost + path;
        }
        var suffix = ingestionDomain + path;
        switch (target) {
            case 'assets':
                return "https://".concat(this.region, "-assets.").concat(suffix);
            case 'api':
                return "https://".concat(this.region, ".").concat(suffix);
        }
    };
    RequestRouter.prototype.loadScript = function (scriptUrlToLoad, callback) {
        if (this.instance.config.disable_external_dependency_loading) {
            logger.warn("".concat(scriptUrlToLoad, " was requested but loading of external scripts is disabled."));
            return callback('Loading of external scripts is disabled');
        }
        var url = scriptUrlToLoad[0] === '/' ? this.endpointFor('assets', scriptUrlToLoad) : scriptUrlToLoad;
        var addScript = function () {
            var _a;
            if (!document) {
                return callback('document not found');
            }
            var scriptTag = document.createElement('script');
            scriptTag.type = 'text/javascript';
            scriptTag.src = url;
            scriptTag.onload = function (event) { return callback(undefined, event); };
            scriptTag.onerror = function (error) { return callback(error); };
            var scripts = document.querySelectorAll('body > script');
            if (scripts.length > 0) {
                (_a = scripts[0].parentNode) === null || _a === void 0 ? void 0 : _a.insertBefore(scriptTag, scripts[0]);
            }
            else {
                // In exceptional situations this call might load before the DOM is fully ready.
                document.body.appendChild(scriptTag);
            }
        };
        if (document === null || document === void 0 ? void 0 : document.body) {
            addScript();
        }
        else {
            document === null || document === void 0 ? void 0 : document.addEventListener('DOMContentLoaded', addScript);
        }
    };
    return RequestRouter;
}());
export { RequestRouter };
